src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -352,60 +404,501 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+
+	/ Cache block locally /
+	ret = source->driver->get_cache_driver()->put(dpp, source->get_key().get_oid(), bl, len, source->get_attrs()); // May be put_async -Sam
+
+	if (!ret) {
+	  int updateValueReturn = source->driver->get_block_dir()->update_field(source->driver->get_cache_block(), "hostsList", ""/local cache ip from config/);
+
+	  if (updateValueReturn < 0) {
+	    ldpp_dout(dpp, 20) << "D4N Filter: Block directory update value operation failed." << dendl;
+	  } else {
+	    ldpp_dout(dpp, 20) << "D4N Filter: Block directory update value operation succeeded." << dendl;
+	  }
+	  
+	  cb->handle_data(bl, ofs, len);
+	}
+      } else {
+	/ Backend store retrieval /
+	ldpp_dout(dpp, 20) << "D4N Filter: Object directory get operation failed." << dendl;
+	ret = next->iterate(dpp, ofs, end, cb, y);
+
+	if (!ret) {
+	  / Cache block locally /
+	  ret = source->driver->get_cache_driver()->put(dpp, source->get_key().get_oid(), bl, len, source->get_attrs()); // May be put_async -Sam
+
+	  / Store block in directory /
+	  rgw::d4n::BlockDirectory* tempBlockDir = source->driver->get_block_dir(); // remove later -Sam
+
+	  source->driver->get_cache_block()->hostsList.push_back(tempBlockDir->get_addr().host + ":" + std::to_string(tempBlockDir->get_addr().port)); // local cache address -Sam 
+	  source->driver->get_cache_block()->size = source->get_obj_size();
+	  source->driver->get_cache_block()->cacheObj.bucketName = source->get_bucket()->get_name();
+	  source->driver->get_cache_block()->cacheObj.objName = source->get_key().get_oid();
+
+	  int setDirReturn = tempBlockDir->set_value(source->driver->get_cache_block());
+
+	  if (setDirReturn < 0) {
+	    ldpp_dout(dpp, 20) << "D4N Filter: Block directory set operation failed." << dendl;
+	  } else {
+	    ldpp_dout(dpp, 20) << "D4N Filter: Block directory set operation succeeded." << dendl;
+	  }
+	}
+      }
+    }
+  }
+
+  if (ret < 0) 
+    ldpp_dout(dpp, 20) << "D4N Filter: Cache iterate operation failed." << dendl;
+
+  return next->iterate(dpp, ofs, end, cb, y); */
+}
+
+int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::flush_last_part()
+{
+  last_part = true;
+  return handle_data(bl_rem, 0, bl_rem.length());
+}
+
+int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl, off_t bl_ofs, off_t bl_len)
+{
+  auto rgw_get_obj_max_req_size = g_conf()->rgw_get_obj_max_req_size;
+
+  if (!last_part && bl.length() <= rgw_get_obj_max_req_size) {
+    auto r = client_cb->handle_data(bl, bl_ofs, bl_len);
+
+    if (r < 0) {
+      return r;
+    }
+  }
+
+  //Accumulating data from backend store into rgw_get_obj_max_req_size sized chunks and then writing to cache
+  if (write_to_cache) {
+    const std::lock_guard l(d4n_get_data_lock);
+    rgw::d4n::CacheBlock block;
+    rgw::d4n::BlockDirectory* blockDir = source->driver->get_block_dir();
+    block.hostsList.push_back(blockDir->get_addr().host + ":" + std::to_string(blockDir->get_addr().port));
```

Comment:
Should the entire object also be stored in the local cache or only the block?

Reply:
Additionally, the entire block must be filled out correctly by adding accurate values for the creation time, version, and dirty bit.

Reply:
We are storing blocks in case of a read cache. When we have the remote caches plugged, all the blocks will be distributed to different caches available. Currently we are caching all blocks belonging to the same object in the same cache. If we read the whole object from the backend, we will break it up into blocks and store them in the cache. If we read partial data from the backend (in case of ranged requests), we will save that in the cache

Reply:
> Additionally, the entire block must be filled out correctly by adding accurate values for the creation time, version, and dirty bit. I will work on this but the version is still not set up

Correct, I have added whatever I could. You are free to fill the other variables - I can see that you have already done that in your PR. I am not sure about the `dirty bit` though, since that applies to the `write-back` cache, which will have data to be written to the backend store.

Reply:
I understand that the blocks that are chunked will be stored in the cache, but I am wondering if I should also update the cache object's host list to store the local cache address. I assume I shouldn't do this because `put_async` is only being called on the block and not the entire object. The object will be handled separately elsewhere, correct?

Reply:
In regards to the `dirty` bit, I'm also not certain how that will be used but since the write and read caches are being unified, I will keep it for the write cache to use. I can also ask how this should be handled in the read workflow, if at all. Do we plan to have the `version` feature in this PR? If so, I can ask Casey more on implementation details.

Reply:
Update: the `dirty` bit will not be used in the read workflow, so it will simply have a default value of `false` when being set here.